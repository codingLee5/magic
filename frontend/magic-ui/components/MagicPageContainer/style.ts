import { createStyles } from "antd-style"

export const useStyles = createStyles(({ css, isDarkMode, prefixCls, token }) => {
	return {
		card: {
			height: "100%",
			overflow: "hidden",
			borderRadius: 0,
			display: "flex",
			flexDirection: "column",
			backgroundColor: `${isDarkMode ? "#141414" : token.magicColorUsages.white}`,
			[`.${prefixCls}-card-body`]: {
				flex: 1,
				display: "flex",
			},
		},
		cardHeader: {
			borderRadius: "0 !important",
			position: "sticky",
			top: 0,
			color: `${
				isDarkMode ? token.magicColorScales.grey[8] : token.magicColorUsages.text[1]
			} !important`,
			zIndex: 10,
			backdropFilter: "blur(12px)",
			background: isDarkMode
				? `${token.magicColorScales.grey[0]} !important`
				: token.magicColorUsages.white,
			[`--${prefixCls}-padding-lg`]: "20px",
		},
		cardBody: css`
			height: calc(100vh - 55px);
			--${prefixCls}-padding-lg: 0;
      --${prefixCls}-card-body-padding: 0;
		`,

		closeButton: css`
			.${prefixCls}-btn-icon {
				height: 24px;
			}
		`,
	}
})
