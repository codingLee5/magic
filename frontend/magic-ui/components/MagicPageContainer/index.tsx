import { Card, Flex, type CardProps } from "antd"
import type { ReactNode } from "react"
import { forwardRef, memo, useMemo } from "react"
import { IconX } from "@tabler/icons-react"
import MagicButton from "../MagicButton"
import MagicIcon from "../MagicIcon"
import { useStyles } from "./style"

export interface MagicPageContainerProps extends CardProps {
	className?: string
	icon?: ReactNode
	closeable?: boolean
	onClose?: () => void
}

const PageContainer = memo(
	forwardRef<HTMLDivElement, MagicPageContainerProps>(
		(
			{
				children,
				className,
				classNames,
				icon,
				title: _title,
				closeable = false,
				onClose,
				...props
			},
			ref,
		) => {
			const { styles, cx } = useStyles()

			const title = useMemo(() => {
				return (
					<Flex align="center" justify="space-between">
						<Flex align="center" gap={8}>
							{icon}
							{_title}
						</Flex>
						<MagicButton
							type="text"
							hidden={!closeable}
							className={styles.closeButton}
							icon={<MagicIcon component={IconX} />}
							onClick={onClose}
						/>
					</Flex>
				)
			}, [_title, closeable, icon, onClose, styles.closeButton])

			return (
				<Card
					ref={ref}
					className={cx(styles.card, className)}
					variant="borderless"
					classNames={{
						header: cx(styles.cardHeader, classNames?.header),
						body: cx(styles.cardBody, classNames?.body),
					}}
					title={title}
					{...props}
				>
					{children}
				</Card>
			)
		},
	),
)

export default PageContainer
