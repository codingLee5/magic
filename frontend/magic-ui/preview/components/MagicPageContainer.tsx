import React from "react"
import MagicPageContainer from "../../components/MagicPageContainer"
import ComponentDemo from "./Container"
import { Button } from "antd"

const PageContainerDemo: React.FC = () => {
	return (
		<div>
			<ComponentDemo
				title="基础页面容器"
				description="最基本的页面容器组件"
				code="<MagicPageContainer title='页面标题'><div>内容</div></MagicPageContainer>"
			>
				<MagicPageContainer title="页面标题">
					<p>这是页面容器的内容区域。</p>
					<p>可以包含任何React组件和内容。</p>
				</MagicPageContainer>
			</ComponentDemo>

			<ComponentDemo title="带副标题" description="支持副标题的页面容器" code="subTitle">
				<MagicPageContainer title="页面标题" extra={<div>页面副标题</div>}>
					<p>这是带副标题的页面容器。</p>
				</MagicPageContainer>
			</ComponentDemo>

			<ComponentDemo title="带标签页" description="支持标签页的页面容器" code="tabList">
				<MagicPageContainer
					title="页面标题"
					tabList={[
						{ key: "tab1", tab: "标签1" },
						{ key: "tab2", tab: "标签2" },
						{ key: "tab3", tab: "标签3" },
					]}
				>
					<p>这是带标签页的页面容器。</p>
				</MagicPageContainer>
			</ComponentDemo>

			<ComponentDemo title="带操作按钮" description="支持操作按钮的页面容器" code="extra">
				<MagicPageContainer
					title="页面标题"
					extra={[
						<Button key="1" style={{ marginRight: 8 }}>
							操作1
						</Button>,
						<Button key="2" type="primary">
							操作2
						</Button>,
					]}
				>
					<p>这是带操作按钮的页面容器。</p>
				</MagicPageContainer>
			</ComponentDemo>
		</div>
	)
}

export default PageContainerDemo
