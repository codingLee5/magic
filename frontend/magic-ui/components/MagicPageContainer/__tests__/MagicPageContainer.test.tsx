import { render, screen } from "@testing-library/react"
import MagicThemeProvider from "../../ThemeProvider"
import { describe, it, expect } from "vitest"
import MagicPageContainer from "../index"

const renderWithTheme = (component: React.ReactElement) =>
	render(<MagicThemeProvider theme="light">{component}</MagicThemeProvider>)

describe("MagicPageContainer", () => {
	it("应该正常渲染", () => {
		renderWithTheme(<MagicPageContainer title="页面标题">页面内容</MagicPageContainer>)
		expect(screen.getByText("页面标题")).toBeInTheDocument()
		expect(screen.getByText("页面内容")).toBeInTheDocument()
	})

	// 快照测试
	describe("快照测试", () => {
		it("基础页面容器快照", () => {
			const { asFragment } = renderWithTheme(
				<MagicPageContainer title="页面标题">页面内容</MagicPageContainer>,
			)
			expect(asFragment()).toMatchSnapshot()
		})

		it("带类名页面容器快照", () => {
			const { asFragment } = renderWithTheme(
				<MagicPageContainer title="页面标题" className="custom-page">
					页面内容
				</MagicPageContainer>,
			)
			expect(asFragment()).toMatchSnapshot()
		})

		it("带自定义属性页面容器快照", () => {
			const { asFragment } = renderWithTheme(
				<MagicPageContainer title="页面标题" data-testid="page-container">
					页面内容
				</MagicPageContainer>,
			)
			expect(asFragment()).toMatchSnapshot()
		})

		it("带操作按钮页面容器快照", () => {
			const { asFragment } = renderWithTheme(
				<MagicPageContainer title="页面标题" extra={<button>操作按钮</button>}>
					页面内容
				</MagicPageContainer>,
			)
			expect(asFragment()).toMatchSnapshot()
		})

		it("带标签页页面容器快照", () => {
			const { asFragment } = renderWithTheme(
				<MagicPageContainer
					title="页面标题"
					tabList={[
						{ key: "tab1", tab: "标签1" },
						{ key: "tab2", tab: "标签2" },
					]}
				>
					页面内容
				</MagicPageContainer>,
			)
			expect(asFragment()).toMatchSnapshot()
		})

		it("带自定义样式页面容器快照", () => {
			const { asFragment } = renderWithTheme(
				<MagicPageContainer title="页面标题" style={{ backgroundColor: "#f0f0f0" }}>
					页面内容
				</MagicPageContainer>,
			)
			expect(asFragment()).toMatchSnapshot()
		})

		it("无标题页面容器快照", () => {
			const { asFragment } = renderWithTheme(
				<MagicPageContainer>页面内容</MagicPageContainer>,
			)
			expect(asFragment()).toMatchSnapshot()
		})
	})
})
