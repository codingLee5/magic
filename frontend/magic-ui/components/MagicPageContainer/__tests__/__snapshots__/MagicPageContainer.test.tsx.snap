// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MagicPageContainer > 快照测试 > 基础页面容器快照 1`] = `
<DocumentFragment>
  <div
    class="magic-card acss-1t0i3xo css-dev-only-do-not-override-yotw4n css-var-r1"
  >
    <div
      class="magic-card-head acss-3n96tu"
    >
      <div
        class="magic-card-head-wrapper"
      >
        <div
          class="magic-card-head-title"
        >
          <div
            class="magic-flex css-dev-only-do-not-override-yotw4n css-var-r1 magic-flex-align-center magic-flex-justify-space-between"
          >
            <div
              class="magic-flex css-dev-only-do-not-override-yotw4n css-var-r1 magic-flex-align-center"
              style="gap: 8px;"
            >
              页面标题
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="magic-card-body acss-jnieq4"
    >
      页面内容
    </div>
  </div>
</DocumentFragment>
`;

exports[`MagicPageContainer > 快照测试 > 带操作按钮页面容器快照 1`] = `
<DocumentFragment>
  <div
    class="magic-card acss-1t0i3xo css-dev-only-do-not-override-yotw4n css-var-r4"
  >
    <div
      class="magic-card-head acss-3n96tu"
    >
      <div
        class="magic-card-head-wrapper"
      >
        <div
          class="magic-card-head-title"
        >
          <div
            class="magic-flex css-dev-only-do-not-override-yotw4n css-var-r4 magic-flex-align-center magic-flex-justify-space-between"
          >
            <div
              class="magic-flex css-dev-only-do-not-override-yotw4n css-var-r4 magic-flex-align-center"
              style="gap: 8px;"
            >
              页面标题
            </div>
          </div>
        </div>
        <div
          class="magic-card-extra"
        >
          <button>
            操作按钮
          </button>
        </div>
      </div>
    </div>
    <div
      class="magic-card-body acss-jnieq4"
    >
      页面内容
    </div>
  </div>
</DocumentFragment>
`;

exports[`MagicPageContainer > 快照测试 > 带标签页页面容器快照 1`] = `
<DocumentFragment>
  <div
    class="magic-card magic-card-contain-tabs acss-1t0i3xo css-dev-only-do-not-override-yotw4n css-var-r5"
  >
    <div
      class="magic-card-head acss-3n96tu"
    >
      <div
        class="magic-card-head-wrapper"
      >
        <div
          class="magic-card-head-title"
        >
          <div
            class="magic-flex css-dev-only-do-not-override-yotw4n css-var-r5 magic-flex-align-center magic-flex-justify-space-between"
          >
            <div
              class="magic-flex css-dev-only-do-not-override-yotw4n css-var-r5 magic-flex-align-center"
              style="gap: 8px;"
            >
              页面标题
            </div>
          </div>
        </div>
      </div>
      <div
        class="magic-tabs magic-tabs-top magic-tabs-large magic-card-head-tabs css-dev-only-do-not-override-yotw4n css-var-r5 magic-tabs-css-var"
      >
        <div
          aria-orientation="horizontal"
          class="magic-tabs-nav"
          role="tablist"
        >
          <div
            class="magic-tabs-nav-wrap"
          >
            <div
              class="magic-tabs-nav-list"
              style="transform: translate(0px, 0px);"
            >
              <div
                class="magic-tabs-tab magic-tabs-tab-active"
                data-node-key="tab1"
              >
                <div
                  aria-controls="rc-tabs-test-panel-tab1"
                  aria-selected="true"
                  class="magic-tabs-tab-btn"
                  id="rc-tabs-test-tab-tab1"
                  role="tab"
                  tabindex="0"
                >
                  标签1
                </div>
              </div>
              <div
                class="magic-tabs-tab"
                data-node-key="tab2"
              >
                <div
                  aria-controls="rc-tabs-test-panel-tab2"
                  aria-selected="false"
                  class="magic-tabs-tab-btn"
                  id="rc-tabs-test-tab-tab2"
                  role="tab"
                  tabindex="-1"
                >
                  标签2
                </div>
              </div>
              <div
                class="magic-tabs-ink-bar magic-tabs-ink-bar-animated"
              />
            </div>
          </div>
          <div
            class="magic-tabs-nav-operations magic-tabs-nav-operations-hidden"
          >
            <button
              aria-controls="rc-tabs-test-more-popup"
              aria-expanded="false"
              aria-haspopup="listbox"
              class="magic-tabs-nav-more"
              id="rc-tabs-test-more"
              style="visibility: hidden; order: 1;"
              type="button"
            >
              <span
                aria-label="ellipsis"
                class="anticon anticon-ellipsis"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="ellipsis"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M176 511a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0z"
                  />
                </svg>
              </span>
            </button>
          </div>
        </div>
        <div
          class="magic-tabs-content-holder"
        >
          <div
            class="magic-tabs-content magic-tabs-content-top"
          >
            <div
              aria-hidden="false"
              aria-labelledby="rc-tabs-test-tab-tab1"
              class="magic-tabs-tabpane magic-tabs-tabpane-active"
              id="rc-tabs-test-panel-tab1"
              role="tabpanel"
              tabindex="0"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="magic-card-body acss-jnieq4"
    >
      页面内容
    </div>
  </div>
</DocumentFragment>
`;

exports[`MagicPageContainer > 快照测试 > 带类名页面容器快照 1`] = `
<DocumentFragment>
  <div
    class="magic-card acss-1t0i3xo custom-page css-dev-only-do-not-override-yotw4n css-var-r2"
  >
    <div
      class="magic-card-head acss-3n96tu"
    >
      <div
        class="magic-card-head-wrapper"
      >
        <div
          class="magic-card-head-title"
        >
          <div
            class="magic-flex css-dev-only-do-not-override-yotw4n css-var-r2 magic-flex-align-center magic-flex-justify-space-between"
          >
            <div
              class="magic-flex css-dev-only-do-not-override-yotw4n css-var-r2 magic-flex-align-center"
              style="gap: 8px;"
            >
              页面标题
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="magic-card-body acss-jnieq4"
    >
      页面内容
    </div>
  </div>
</DocumentFragment>
`;

exports[`MagicPageContainer > 快照测试 > 带自定义属性页面容器快照 1`] = `
<DocumentFragment>
  <div
    class="magic-card acss-1t0i3xo css-dev-only-do-not-override-yotw4n css-var-r3"
    data-testid="page-container"
  >
    <div
      class="magic-card-head acss-3n96tu"
    >
      <div
        class="magic-card-head-wrapper"
      >
        <div
          class="magic-card-head-title"
        >
          <div
            class="magic-flex css-dev-only-do-not-override-yotw4n css-var-r3 magic-flex-align-center magic-flex-justify-space-between"
          >
            <div
              class="magic-flex css-dev-only-do-not-override-yotw4n css-var-r3 magic-flex-align-center"
              style="gap: 8px;"
            >
              页面标题
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="magic-card-body acss-jnieq4"
    >
      页面内容
    </div>
  </div>
</DocumentFragment>
`;

exports[`MagicPageContainer > 快照测试 > 带自定义样式页面容器快照 1`] = `
<DocumentFragment>
  <div
    class="magic-card acss-1t0i3xo css-dev-only-do-not-override-yotw4n css-var-r7"
    style="background-color: rgb(240, 240, 240);"
  >
    <div
      class="magic-card-head acss-3n96tu"
    >
      <div
        class="magic-card-head-wrapper"
      >
        <div
          class="magic-card-head-title"
        >
          <div
            class="magic-flex css-dev-only-do-not-override-yotw4n css-var-r7 magic-flex-align-center magic-flex-justify-space-between"
          >
            <div
              class="magic-flex css-dev-only-do-not-override-yotw4n css-var-r7 magic-flex-align-center"
              style="gap: 8px;"
            >
              页面标题
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="magic-card-body acss-jnieq4"
    >
      页面内容
    </div>
  </div>
</DocumentFragment>
`;

exports[`MagicPageContainer > 快照测试 > 无标题页面容器快照 1`] = `
<DocumentFragment>
  <div
    class="magic-card acss-1t0i3xo css-dev-only-do-not-override-yotw4n css-var-r8"
  >
    <div
      class="magic-card-head acss-3n96tu"
    >
      <div
        class="magic-card-head-wrapper"
      >
        <div
          class="magic-card-head-title"
        >
          <div
            class="magic-flex css-dev-only-do-not-override-yotw4n css-var-r8 magic-flex-align-center magic-flex-justify-space-between"
          >
            <div
              class="magic-flex css-dev-only-do-not-override-yotw4n css-var-r8 magic-flex-align-center"
              style="gap: 8px;"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="magic-card-body acss-jnieq4"
    >
      页面内容
    </div>
  </div>
</DocumentFragment>
`;
